lib/logstash/outputs/logstash.rb

```diff
@@ -65,125 +85,236 @@ def initialize(*a)
+      fail(LogStash::ConfigurationError, 'SSL trust can be configured with EITHER `ssl_certificate_authorities` OR `ssl_truststore_*`, but not both')
+    elsif @ssl_certificate_authorities&.any?
+      fail(LogStash::ConfigurationError, 'SSL Certificate Authorities cannot be configured when `ssl_verification_mode => none`') if @ssl_verification_mode == 'none'
+    elsif @ssl_truststore_path
+      fail(LogStash::ConfigurationError, 'SSL Truststore cannot be configured when `ssl_verification_mode => none`') if @ssl_verification_mode == 'none'
+      fail(LogStash::ConfigurationError, '`ssl_truststore_password` is REQUIRED when `ssl_truststore_path` is provided') if @ssl_truststore_password.nil?
+
+    elsif @ssl_truststore_password
+      fail(LogStash::ConfigurationError, '`ssl_truststore_password` not allowed unless `ssl_truststore_path` is configured')
+    end
+  end
 
-        trust_options['ssl_certificate_authorities'] = @ssl_certificate_authorities.first
-      elsif @ssl_truststore_path
-        fail(LogStash::ConfigurationError, 'SSL Truststore cannot be configured when `ssl_verification_mode => none`') if @ssl_verification_mode == 'none'
+  def construct_host_uri
+    scheme = @ssl_enabled ? 'https'.freeze : 'http'.freeze
+    [].tap do |downstream_uris|
+      @hosts.each do | host_port_pair | # Struct(:host, :port)
+        uri = LogStash::Util::SafeURI.new(host_port_pair[:host])
+        uri.port = host_port_pair[:port].nil? ? DEFAULT_PORT : host_port_pair[:port]
+        uri.update(:scheme, scheme)
+        # we only need `SafeURI::String` to directly apply to Manticore client
+        downstream_uris << uri.to_s.freeze
+      end
+    end
+  end
 
-        trust_options['ssl_truststore_path'] = @ssl_truststore_path
-        trust_options['ssl_truststore_password'] = @ssl_truststore_password || fail(LogStash::ConfigurationError, '`ssl_truststore_password` is REQUIRED when `ssl_truststore_path` is provided')
-      elsif @ssl_truststore_password
-        fail(LogStash::ConfigurationError, '`ssl_truststore_password` not allowed unless `ssl_truststore_path` is configured')
+  def send_events(events)
+    # we use array to utilize `peek` feature, note that every thread has its own array here
+    pending = Array.new
+    pending << [:send, events, 0]
+
+    while (popped = pending.shift)
+      action, events, attempt = popped
+      break if action == :done
+
+      # this may happen when _all_ upstream inputs are unreachable (ex: timeout) and we are _retrying_ to send events
+      raise PluginInternalQueueLeftoverError.new("Received pipeline shutdown request but Logstash output has unfinished events. " \
+              "If persistent queue is enabled, events will be retried.") if attempt > 2 && pipeline_shutdown_requested?
+
+      # TODO: what if all hosts are unreachable in a given cool_off period?
+      # sleep? or drop the event
+      # also, retry doesn't have any meaning if all hosts are unreachable
```

Comment:
`cool_off` period also doen't make sense if all hosts fail, because route is sorting the hosts based on `last_request_error_time + cool_off` and `last_request_time`.

Talk to team what behavior would be acceptable.